File: script/rdf_build_support.py
from __future__ import annotations

import configparser
import gzip
from pathlib import Path
from typing import Iterable, TextIO


SCRIPT_DIR = Path(__file__).resolve().parent
REPO_ROOT = SCRIPT_DIR.parent

DEFAULT_RESOURCE_ROOTS = {
    "ncbigene_dir": Path("data/NCBIGene"),
    "medgen_dir": Path("data/MedGen"),
    "orphanet_dir": Path("data/Orphanet"),
    "mondo_dir": Path("data/MONDO"),
    "gencc_dir": Path("data/GenCC"),
    "panelsearch_dir": Path("data/PanelSearch"),
    "omim_dir": Path("data/OMIM"),
    "kegg_dir": Path("data/KEGG"),
    "genereviews_dir": Path("data/GeneReviews"),
    "hpo_dir": Path("data/HPO"),
}


def dotted_to_snake(key: str) -> str:
    return key.replace(".", "_")


def load_config(config_path: str | Path | None = None) -> dict[str, str]:
    path = _find_config_path(config_path)
    if path is None:
        return {}

    text = path.read_text(encoding="utf-8")
    parser = configparser.ConfigParser()

    if "[DEFAULT]" in text or any(line.strip().startswith("[") for line in text.splitlines()):
        parser.read_string(text)
        return {key: value for key, value in parser.defaults().items()}

    config: dict[str, str] = {}
    for raw_line in text.splitlines():
        line = raw_line.strip()
        if not line or line.startswith("#") or line.startswith("!"):
            continue
        separator = "=" if "=" in line else ":"
        if separator not in line:
            continue
        key, value = line.split(separator, 1)
        config[key.strip()] = value.strip()
    return config


def _find_config_path(config_path: str | Path | None) -> Path | None:
    if config_path is not None:
        path = Path(config_path)
        return path if path.exists() else None

    candidates = [
        Path("config.ini"),
        REPO_ROOT / "config.ini",
        SCRIPT_DIR / "config.ini",
        Path("pcf-rdf.properties"),
        REPO_ROOT / "pcf-rdf.properties",
        SCRIPT_DIR / "pcf-rdf.properties",
    ]
    for candidate in candidates:
        if candidate.exists():
            return candidate
    return None


def trim_to_none(value: str | None) -> str | None:
    if value is None:
        return None
    stripped = value.strip()
    return stripped if stripped else None


def config_value(config: dict[str, str], *keys: str) -> str | None:
    for key in keys:
        value = trim_to_none(config.get(key))
        if value is not None:
            return value

        snake_key = dotted_to_snake(key)
        value = trim_to_none(config.get(snake_key))
        if value is not None:
            return value
    return None


def normalize_path(value: str | Path) -> str:
    return str(value).replace("\\", "/")


def resolve_configured_output_dir(config: dict[str, str]) -> str:
    configured = config_value(config, "rdf.output.dir", "rdf_output_dir")
    return normalize_path(configured) if configured is not None else "RDF/latest"


def resolve_resource_root(config: dict[str, str], directory_key: str) -> Path:
    configured = config_value(config, directory_key, dotted_to_snake(directory_key))
    if configured is not None:
        return Path(configured)

    snake_key = dotted_to_snake(directory_key)
    if snake_key not in DEFAULT_RESOURCE_ROOTS:
        raise ValueError(f"Unknown directory key: {directory_key}")
    return DEFAULT_RESOURCE_ROOTS[snake_key]


def resolve_configured_file(
    config: dict[str, str],
    exact_path_key: str,
    base_dir: Path,
    file_name: str,
    *,
    alternate_file_names: Iterable[str] = (),
    required: bool = True,
) -> str:
    configured = config_value(config, exact_path_key, dotted_to_snake(exact_path_key))
    if configured is not None:
        return normalize_path(configured)

    candidates = (file_name, *alternate_file_names)
    for candidate_name in candidates:
        resolved = resolve_latest_file(base_dir, candidate_name, required=False)
        if resolved is not None:
            return normalize_path(resolved)

        source_file = resolve_existing_path(
            Path("data/source") / candidate_name,
            REPO_ROOT / "data/source" / candidate_name,
        )
        if source_file.exists():
            return normalize_path(source_file)

    if required:
        names = ", ".join(candidates)
        raise FileNotFoundError(f"Could not find {names} under {base_dir}")
    return ""


def resolve_latest_file(base_dir: Path, file_name: str, *, required: bool = True) -> Path | None:
    base_dir = resolve_existing_path(base_dir, REPO_ROOT / base_dir)
    direct_file = base_dir / file_name
    if direct_file.exists():
        return direct_file

    latest_file = base_dir / "latest" / file_name
    if latest_file.exists():
        return latest_file

    if not base_dir.is_dir():
        if required:
            raise FileNotFoundError(f"Could not find {file_name} under {base_dir}")
        return None

    children = sorted(
        (path for path in base_dir.iterdir() if path.is_dir()),
        key=lambda path: path.name,
        reverse=True,
    )
    for child in children:
        child_file = child / file_name
        if child_file.exists():
            return child_file

    if required:
        raise FileNotFoundError(f"Could not find {file_name} under {base_dir}")
    return None


def resolve_existing_path(*candidates: Path) -> Path:
    for candidate in candidates:
        if candidate.exists():
            return candidate
    return candidates[0]


def open_text_reader(path: str | Path) -> TextIO:
    path = Path(path)
    if path.suffix == ".gz":
        return gzip.open(path, "rt", encoding="utf-8")
    return path.open("rt", encoding="utf-8")


def open_text_writer(path: str | Path) -> TextIO:
    path = Path(path)
    if path.parent:
        path.parent.mkdir(parents=True, exist_ok=True)
    if path.suffix == ".gz":
        return gzip.open(path, "wt", encoding="utf-8", newline="\n")
    return path.open("wt", encoding="utf-8", newline="\n")

File: scripts/java/RdfBuildSupport.java
package pubcasefinder_260415;

	import java.io.BufferedWriter;
	import java.io.BufferedReader;
	import java.io.IOException;
	import java.io.InputStream;
	import java.io.InputStreamReader;
	import java.io.OutputStream;
	import java.io.OutputStreamWriter;
	import java.nio.charset.StandardCharsets;
	import java.nio.file.Files;
	import java.nio.file.Path;
	import java.nio.file.Paths;
	import java.util.LinkedHashMap;
	import java.util.List;
	import java.util.Comparator;
	import java.util.Properties;
	import java.util.stream.Collectors;
	import java.util.stream.Stream;
	import java.util.zip.GZIPInputStream;
	import java.util.zip.GZIPOutputStream;

// RDF 생성기들이 공통으로 사용하는 설정, 경로 해석, gzip 입력 유틸리티
public final class RdfBuildSupport {
	public static final String CONFIG_PATH = "pcf-rdf.properties";

	private RdfBuildSupport() {
	}

	/** 설정 파일을 읽어 경로 override 값을 메모리에 적재한다. */
	public static Properties loadConfig() {
		Properties properties = new Properties();
		Path configPath = Paths.get(CONFIG_PATH);
		if (Files.exists(configPath)) {
			try (InputStream input = Files.newInputStream(configPath)) {
				properties.load(input);
			}
			catch (IOException e) {
				throw new RuntimeException("Failed to load config: " + configPath, e);
			}
		}
		return properties;
	}

	/** 비어 있는 설정 문자열을 null로 바꾼다. */
	public static String trimToNull(String value) {
		if (value == null) {
			return null;
		}
		String trimmed = value.trim();
		return trimmed.isEmpty() ? null : trimmed;
	}

	/** 경로 구분자를 슬래시 기준으로 통일한다. */
	public static String normalizePath(String value) {
		return value.replace('\\', '/');
	}

	/** 설정값 또는 기본값을 사용해 RDF 출력 디렉터리를 결정한다. */
	public static String resolveConfiguredOutputDir(Properties config) {
		String configuredDir = trimToNull(config.getProperty("rdf.output.dir"));
		if (configuredDir != null) {
			return normalizePath(configuredDir);
		}
		return "RDF/latest";
	}

	/** 개별 파일 override 또는 latest 규칙을 이용해 입력 파일 경로를 결정한다. */
	public static String resolveConfiguredFile(Properties config, String exactPathKey, Path baseDir, String fileName) {
		String exactPath = trimToNull(config.getProperty(exactPathKey));
		if (exactPath != null) {
			return normalizePath(exactPath);
		}
		return resolveLatestFile(baseDir, fileName, true).toString().replace('\\', '/');
	}

	/** 디렉터리 키에 대응하는 기본 데이터 루트 경로를 반환한다. */
	public static Path resolveResourceRoot(Properties config, String directoryKey) {
		String configuredDir = trimToNull(config.getProperty(directoryKey));
		if (configuredDir != null) {
			return Paths.get(normalizePath(configuredDir));
		}
		return resolveDefaultResourceRoot(directoryKey);
	}

	private static Path resolveDefaultResourceRoot(String directoryKey) {
		switch (directoryKey) {
		case "ncbigene.dir":
			return Paths.get("data", "NCBIGene");
		case "medgen.dir":
			return Paths.get("data", "MedGen");
		case "orphanet.dir":
			return Paths.get("data", "Orphanet");
		case "mondo.dir":
			return Paths.get("data", "MONDO");
		case "gencc.dir":
			return Paths.get("data", "GenCC");
		case "panelsearch.dir":
			return Paths.get("data", "PanelSearch");
		case "omim.dir":
			return Paths.get("data", "OMIM");
		case "kegg.dir":
			return Paths.get("data", "KEGG");
		case "genereviews.dir":
			return Paths.get("data", "GeneReviews");
		case "hpo.dir":
			return Paths.get("data", "HPO");
		default:
			throw new IllegalArgumentException("Unknown directory key: " + directoryKey);
		}
	}

	/** direct/latest/최신 날짜 폴더 순서로 대상 파일을 탐색한다. */
	public static Path resolveLatestFile(Path baseDir, String fileName, boolean required) {
		Path directFile = baseDir.resolve(fileName);
		if (Files.exists(directFile)) {
			return directFile;
		}

		Path latestFile = baseDir.resolve("latest").resolve(fileName);
		if (Files.exists(latestFile)) {
			return latestFile;
		}

		if (!Files.isDirectory(baseDir)) {
			if (required) {
				throw new IllegalStateException("Could not find " + fileName + " under " + baseDir);
			}
			return null;
		}

		try (Stream<Path> children = Files.list(baseDir)) {
			Path resolved = children
					.filter(Files::isDirectory)
					.sorted(Comparator.comparing((Path path) -> path.getFileName().toString()).reversed())
					.map(path -> path.resolve(fileName))
					.filter(Files::exists)
					.findFirst()
					.orElse(null);
			if (resolved == null && required) {
				throw new IllegalStateException("Could not find " + fileName + " under " + baseDir);
			}
			return resolved;
		}
		catch (IOException e) {
			throw new RuntimeException("Failed to resolve latest file for " + fileName + " under " + baseDir, e);
		}
	}

	/** 파일 확장자에 따라 일반 입력 또는 gzip 입력 스트림을 연다. */
	public static InputStream openMaybeGzip(String path) throws IOException {
		InputStream input = Files.newInputStream(Paths.get(path));
		return path.endsWith(".gz") ? new GZIPInputStream(input) : input;
	}

	/** UTF-8 기준으로 일반/압축 입력 파일을 읽는 BufferedReader를 연다. */
	public static BufferedReader openUtf8Reader(String path) throws IOException {
		return new BufferedReader(new InputStreamReader(openMaybeGzip(path), StandardCharsets.UTF_8));
	}

	/** 출력 파일의 상위 디렉터리가 없으면 생성한다. */
	public static void ensureParentDirectory(String outputPath) throws IOException {
		Path parent = Paths.get(outputPath).getParent();
		if (parent != null) {
			Files.createDirectories(parent);
		}
	}

	/** 파일 확장자에 따라 UTF-8 일반 출력 또는 UTF-8 gzip 출력을 연다. */
	public static BufferedWriter createBufferedWriter(String outputPath) throws IOException {
		ensureParentDirectory(outputPath);
		OutputStream output = Files.newOutputStream(Paths.get(outputPath));
		if (outputPath.endsWith(".gz")) {
			output = new GZIPOutputStream(output);
		}
		return new BufferedWriter(new OutputStreamWriter(output, StandardCharsets.UTF_8));
	}

	/** 설정값, 번들 tools 폴더, 얕은 파일 탐색 순서로 실행 파일 경로를 결정한다. */
	public static String resolveExecutablePath(Properties config, String configKey, String fileName) throws IOException {
		String configured = trimToNull(config.getProperty(configKey));
		if (configured != null && Files.exists(Paths.get(configured))) {
			return Paths.get(configured).toString();
		}

		Path bundledTool = Paths.get("tools", "ncbi", fileName);
		if (Files.exists(bundledTool)) {
			return bundledTool.toAbsolutePath().normalize().toString();
		}

		List<Path> matches;
		try (Stream<Path> paths = Files.walk(Paths.get("."), 4)) {
			matches = paths
					.filter(path -> Files.isRegularFile(path) && path.getFileName().toString().equalsIgnoreCase(fileName))
					.collect(Collectors.toList());
		}

		if (matches.isEmpty()) {
			throw new IOException("Could not find " + fileName + ". Set " + configKey + " in " + CONFIG_PATH + " if needed.");
		}
		return matches.get(0).toAbsolutePath().normalize().toString();
	}

	/** 문자열 쌍을 순서가 보존되는 맵으로 만든다. */
	public static LinkedHashMap<String, String> createStringMap(String... keyValuePairs) {
		if (keyValuePairs.length % 2 != 0) {
			throw new IllegalArgumentException("Key/value pairs must be even.");
		}

		LinkedHashMap<String, String> map = new LinkedHashMap<String, String>();
		for (int i = 0; i < keyValuePairs.length; i += 2) {
			map.put(keyValuePairs[i], keyValuePairs[i + 1]);
		}
		return map;
	}
}
